File: app/src/main/java/com/machinerychorus/lifeprogresswallpaper/DatePickerPreference.java
package com.machinerychorus.lifeprogresswallpaper;

import android.content.Context;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.DatePicker;

import org.joda.time.LocalDate;

/**
 * Date picker preference. This is implemented poorly because I am lazy; It saves the date as a string
 * Created by Ethan on 8/13/2017.
 */

public class DatePickerPreference extends DialogPreference {
    private DatePicker picker;
    private LocalDate lastDate;

    public DatePickerPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public DatePickerPreference(Context context, AttributeSet attrs, int defStyle){
        super(context, attrs, defStyle);
    }

    @Override
    protected View onCreateDialogView() {
        picker=new DatePicker(getContext());

        return(picker);
    }

    @Override
    protected void onBindDialogView(View v) {
        super.onBindDialogView(v);

        if(lastDate != null) {
            //need to -1/+1 the month because the picker uses 0-11 and jodatime uses 1-12
            picker.updateDate(lastDate.getYear(), lastDate.getMonthOfYear()-1, lastDate.getDayOfMonth());
        }
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);

        if (positiveResult) {
            lastDate = new LocalDate(picker.getYear(), picker.getMonth()+1, picker.getDayOfMonth());
            persistString(lastDate.toString());
        }
    }

    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
        lastDate=(restoreValue ? LocalDate.parse(getPersistedString("1994")) : new LocalDate());
    }
}

File: app/src/main/java/com/machinerychorus/lifeprogresswallpaper/DatePickerPreference.kt
package com.machinerychorus.lifeprogresswallpaper

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.util.AttributeSet
import android.view.View
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import androidx.preference.DialogPreference
import org.joda.time.LocalDate
import java.util.*

/**
 * Date picker preference. This is implemented poorly because I am lazy; It saves the date as a string
 * Created by Ethan on 8/13/2017.
 * WIP: fixing for androidx https://stackoverflow.com/questions/52754655/ what a pita
 */
class DatePickerPreference(context: Context?, attrs: AttributeSet?) :
    DialogPreference(context, attrs) {
    private var lastDate: LocalDate? = null

    //TODO: needs more work for androidx migration
    protected fun onBindDialogView(v: View?) {
        super.onBindDialogView(v)
        if (lastDate != null) {
            //need to -1/+1 the month because the picker uses 0-11 and jodatime uses 1-12
            picker.updateDate(lastDate!!.year, lastDate!!.monthOfYear - 1, lastDate!!.dayOfMonth)
        }
    }

    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any) {
        lastDate = if (restoreValue) LocalDate.parse(getPersistedString("1994")) else LocalDate()
    }

    class DatePickerFragment : DialogFragment(), OnDateSetListener {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            // Use the current date as the default date in the picker
            val c = Calendar.getInstance()
            val year = c[Calendar.YEAR]
            val month = c[Calendar.MONTH]
            val day = c[Calendar.DAY_OF_MONTH]

            // Create a new instance of DatePickerDialog and return it
            return DatePickerDialog(activity!!, this, year, month, day)
        }

        override fun onDateSet(view: DatePicker, year: Int, month: Int, day: Int) {
            // Do something with the date chosen by the user
        }
    }
}